test/test_disjoint_set.py

```diff
@@ -121,3 +121,28 @@ def test_itersets_iterates_over_all_component_classes(self):
             list(self.dset.itersets()),
             [{1, 2, 3}, {4, 5}],
         )
+
+    def test_deep_recursion(self):
```

Comment:
I think this test is a solid proof that your approach works, but I'm not sure if we need to keep this in the codebase - if we get rid of the recursive approach, then having a test that ensures our logic doesn't exceed the recursion limit is kind of redundant.